consensus/hotstuff/voteaggregator/vote_aggregator_v2_test.go

```diff
@@ -0,0 +1,251 @@
+	suite.Suite
+
+	mockedCollectors map[uint64]*mocks.VoteCollector
+
+	committee  *mocks.Committee
+	signer     *mocks.SignerVerifier
+	notifier   *mocks.Consumer
+	collectors *mocks.VoteCollectors
+	aggregator *VoteAggregatorV2
+}
+
+func (s *VoteAggregatorV2TestSuite) SetupTest() {
+	s.committee = &mocks.Committee{}
+	s.signer = &mocks.SignerVerifier{}
+	s.notifier = &mocks.Consumer{}
+	s.collectors = &mocks.VoteCollectors{}
+
+	s.mockedCollectors = make(map[uint64]*mocks.VoteCollector)
+
+	s.collectors.On("GetOrCreateCollector", mock.Anything).Return(
+		func(view uint64) hotstuff.VoteCollector {
+			return s.mockedCollectors[view]
+		}, func(view uint64) bool {
+			return true
+		}, func(view uint64) error {
+			_, ok := s.mockedCollectors[view]
+			if !ok {
+				return voteCollectorsFactoryError
+			}
+			return nil
+		})
+
+	var err error
+
+	s.aggregator, err = NewVoteAggregatorV2(unittest.Logger(), s.notifier, 0, s.collectors)
+	require.NoError(s.T(), err)
+
+	// startup aggregator
+	<-s.aggregator.Ready()
+}
+
+func (s *VoteAggregatorV2TestSuite) TearDownTest() {
+	<-s.aggregator.Done()
+}
+
+// prepareMockedCollector prepares a mocked collector and stores it in map, later it will be used
+// to mock behavior of vote collectors.
+func (s *VoteAggregatorV2TestSuite) prepareMockedCollector(view uint64) *mocks.VoteCollector {
+	collector := &mocks.VoteCollector{}
+	collector.On("View").Return(view).Maybe()
+	s.mockedCollectors[view] = collector
+	return collector
+}
+
+// TestAddVote_DeliveryOfQueuedVotes tests if votes are being delivered for processing by queening logic
+func (s *VoteAggregatorV2TestSuite) TestAddVote_DeliveryOfQueuedVotes() {
+	view := uint64(1000)
+	clr := s.prepareMockedCollector(view)
+	votes := make([]*model.Vote, 10)
+	for i := range votes {
+		votes[i] = unittest.VoteFixture(unittest.WithVoteView(view))
+		clr.On("AddVote", votes[i]).Once().Return(nil)
+	}
+
+	for _, vote := range votes {
+		err := s.aggregator.AddVote(vote)
+		require.NoError(s.T(), err)
+	}
+
+	time.Sleep(time.Millisecond * 100)
+
+	clr.AssertExpectations(s.T())
+}
+
+// TestAddVote_StaleVote tests that stale votes(with view lower or equal to highest pruned view) are dropped
+// and not processed.
+func (s *VoteAggregatorV2TestSuite) TestAddVote_StaleVote() {
+	view := uint64(1000)
+	s.collectors.On("PruneUpToView", view).Return(nil)
+	s.aggregator.PruneUpToView(view)
+
+	// try adding votes with view lower than highest pruned
+	votes := 10
+	for i := 0; i < votes; i++ {
+		vote := unittest.VoteFixture(unittest.WithVoteView(view))
+		err := s.aggregator.AddVote(vote)
+		require.NoError(s.T(), err)
+	}
+
+	time.Sleep(time.Millisecond * 100)
+	s.collectors.AssertNotCalled(s.T(), "GetOrCreateCollector")
+}
+
+// TestAddBlock_ValidProposal tests that happy path processing of valid proposal finishes without errors
+func (s *VoteAggregatorV2TestSuite) TestAddBlock_ValidProposal() {
+	view := uint64(1000)
+	clr := s.prepareMockedCollector(view)
```

Comment:
I'm not a big fan of mocking intermediate state for testing like this.

To me, it's best to mock the state transition behavior, and create intermediate states from an initial state using the exposes methods to be tested, rather than creating an intermediate state directly.

In this case here, the vote collectors' state are in-memory state, all possible states (having or not having a certain vote collection) are arrived by only making method calls to the Vote aggregator. 

In a happy path, when calling AddBlock, there are two cases:
1) AddBlock is called before any AddVote was called
2) AddBlock is called after some AddVote was called

So to test the `AddBlock` method, we need to create the above two cases. And we know the difference of the two cases are:
1) no vote collector for that view is created
2) a vote collector for that view has been created

However, instead of creating the vote collector directly, I would suggest to use the `AddVote` to trigger the creation of the vote collector. 

To do that, we need to simulate the `VoteCollectors.GetOrCreateVoteCollector`. It's not hard to implement a thread-safe map that return the vote collector if exists, and create if not.

That way we can setup the case 1) without doing anything, and setup the case 2) by calling `AddVote` first, which will trigger the mocked `VoteCollectors.GetOrCreateVoteCollector` to create the vote collector mock. And it's easier to asset the state of the map in the mocked VoteCollectors at the end of the tests.

So the happy cases can be described as:
1) When AddBlock is called, if there is no vote was added before, there should be only one vote collector created in the map, and the vote collector should be called with `ProcessBlock`
2) AddVote for a block, then calling AddBlock with the block, there should be only one vote collector created in the map, and the vote collector should be called with both `AddVote` and `ProcessBlock`.
3) When calling AddBlock concurrently with 10 different blocks with different views, there should be 10 vote collectors create in the map (this is why the thread-safe vote collector implementation is useful here)
4) When calling AddBlock concurrently with 10 different blocks with the view, there should be only one vote collectors create in the map, and 9 of the 10 calls should be return double proposal error.

Reply:
@zhangchiqing I don't understand the benefit of implementing mock creation this way. 
With current approach we place a hard requirement on logic of `VoteAggregator`, in other words it must call `GetOrCreateCollector` exactly with parameters we are expecting, otherwise it's a failed test. I don't see any harm in _preparing_ for certain call and setting up expected behavior. Of course we can implement a factory and then check if only one object was created, but it's harder for to mock collectors that way + I don't see any benefits to this approach. 
Sorry, maybe I don't understand something :)

Reply:
> In a happy path, when calling AddBlock, there are two cases:
>  1. AddBlock is called before any AddVote was called
>  2. AddBlock is called after some AddVote was called
> So to test the AddBlock method, we need to create the above two cases. 

To me, it is unclear what benefit we gain from differentiating these two conditions. In my view, differentiating the two cases is completely encapsulated in the `VoteCollectors`, i.e. a different component than the one we test here. 
* When receiving a vote, the `VoteAggregator` only has to forward the vote to the respective collector. (whether or not we previously received the block is completely irrelevant for the  `VoteAggregator`)
* When receiving a block, the `VoteAggregator` only has to forward the block to the respective collector. This is also completely independent of any votes. 

I think for the unit test of the `VoteAggregator`, it is completely fine to mock out `VoteCollectors`. 

Nevertheless, I think there is a valid argument to testing `VoteAggregator` in conjunction with `VoteCollectors`:
* with our unit tests, we verify that `VoteAggregator` and `VoteCollectors` independently adhere to their respective API specification
* In our architecture document, we have mathematical reasoning along the line:
   *  If `VoteAggregator` and `VoteCollectors` independently adhere to their respective API specification, then the combined system will be correct
* However, there is a gap between the code-level API specification and the mathematical API specification of `VoteAggregator` and `VoteCollectors`. We are humans and we make mistakes. So there is value in testing that the full implementations of `VoteAggregator` and `VoteCollectors` _together_ work correctly. 
* To me, this is on the level of an integration test, because we are testing multiple components in conjunction with each other.